Distribute ios detect debug

TroubleMakerBen opened the pull request.

msftclas commented:
@TroubleMakerBen,
Thanks for your contribution as a Microsoft full-time employee or intern. You do not need to sign a CLA.
_Thanks,
Microsoft Pull Request Bot_

guperrot commented on SDK/MobileCenter/Microsoft.Azure.Mobile.Android/MobileCenter.cs:
```diff
@@ -184,7 +184,19 @@ private static Application SetWrapperSdkAndGetApplication()
         private static Class[] GetServices(IEnumerable<Type> services)
         {
-            return services.Select(service => Class.FromType((Type)service.GetProperty("BindingType").GetValue(null, null))).ToArray();
+            List<Class> classes = new List<Class>();
```
we always use var instead of type for local variables (resharper convention)

guperrot commented on SDK/MobileCenter/Microsoft.Azure.Mobile.Android/MobileCenter.cs:
```diff
@@ -184,7 +184,19 @@ private static Application SetWrapperSdkAndGetApplication()
+            List<Class> classes = new List<Class>();
+            foreach (Type t in services)
+            {
+                if ((t.GetProperty("BindingType") != null) && ((Type)t.GetProperty("BindingType").GetValue(null, null) != null))
```
please assign t.GetProperty to a local variable and reuse it

guperrot commented on SDK/MobileCenter/Microsoft.Azure.Mobile.iOS/MobileCenter.cs:
```diff
@@ -171,7 +171,16 @@ public static bool Enabled
         private static Class[] GetServices(IEnumerable<Type> services)
         {
-            return services.Select(service => GetClassForType(GetBindingType(service))).ToArray();
+            List<Class> classes = new List<Class>();
```
same comments as in ios

guperrot commented on SDK/MobileCenterDistribute/Microsoft.Azure.Mobile.Distribute.iOS/Distribute.cs:
```diff
@@ -8,7 +8,19 @@ namespace Microsoft.Azure.Mobile.Distribute
+            {
+
+// We cannot use Conditional Attributes as they only work for non-void methods and classes, not for properties.
+#if DEBUG
```
This will not work on consumer side (once in nuget), the sdk is always shipped in release, same issue as ios native.

When you build locally forms puppet you build both sdk and app to debug that's why it worked for you, but in demo debug with nuget it will not work.

guperrot replied:
To fix I suggest having an additional field variable to hold the BindingType value to return in property, and have a conditional void method to set it to null in debug or something like that.

guperrot commented on SDK/MobileCenterDistribute/Microsoft.Azure.Mobile.Distribute.iOS/Distribute.cs:
```diff
@@ -8,7 +8,19 @@ namespace Microsoft.Azure.Mobile.Distribute
+            get
+            {
+
+// We cannot use Conditional Attributes as they only work for non-void methods and classes, not for properties.
```
You meant void instead of non-void?
In Crashes we have a conditional void method that works only in debug (generateTestCrashes).

guperrot commented on SDK/MobileCenterDistribute/Microsoft.Azure.Mobile.Distribute.iOS/Distribute.cs:
```diff
@@ -7,8 +7,24 @@ namespace Microsoft.Azure.Mobile.Distribute
+        static Type _internalBindingType = typeof(iOSDistribute);
+
+        [System.Diagnostics.Conditional("DEBUG")]
+        static void setInternalBindingTypeForDebug()
```
Upper case s at beginning, its c# ;)

guperrot commented on SDK/MobileCenter/Microsoft.Azure.Mobile.iOS/MobileCenter.cs:
```diff
@@ -171,7 +171,16 @@ public static bool Enabled
+            var classes = new List<Class>();
+            foreach (Type t in services)
+            {
+                if((GetBindingType(t) != null) && (GetClassForType(GetBindingType(t)) != null))
```
Use local var for GetBindingType

guperrot commented on SDK/MobileCenter/Microsoft.Azure.Mobile.iOS/MobileCenter.cs:
```diff
@@ -171,7 +171,16 @@ public static bool Enabled
+            {
+                if((GetBindingType(t) != null) && (GetClassForType(GetBindingType(t)) != null))
+                {
+                    Class aClass = GetClassForType(GetBindingType(t));
```
Use var or inline

guperrot commented on SDK/MobileCenterDistribute/Microsoft.Azure.Mobile.Distribute.iOS/Distribute.cs:
```diff
@@ -7,8 +7,24 @@ namespace Microsoft.Azure.Mobile.Distribute
 
     public static partial class Distribute
     {
+
```
remove extra blank line

guperrot commented on SDK/MobileCenter/Microsoft.Azure.Mobile.Android/MobileCenter.cs:
```diff
@@ -184,7 +184,20 @@ private static Application SetWrapperSdkAndGetApplication()
+                var propertyInfo = t.GetProperty("BindingType");
+                if ((propertyInfo != null) && ((Type)propertyInfo.GetValue(null, null) != null))
+                {
+                    Class aClass = Class.FromType((Type)propertyInfo.GetValue(null, null));
```
use var

guperrot commented on SDK/MobileCenter/Microsoft.Azure.Mobile.Android/MobileCenter.cs:
```diff
@@ -184,7 +184,20 @@ private static Application SetWrapperSdkAndGetApplication()
+            foreach (Type t in services)
+            {
+                var propertyInfo = t.GetProperty("BindingType");
+                if ((propertyInfo != null) && ((Type)propertyInfo.GetValue(null, null) != null))
```
Use local var for GetValue return

guperrot commented on SDK/MobileCenter/Microsoft.Azure.Mobile.Android/MobileCenter.cs:
```diff
@@ -188,12 +188,16 @@ private static Class[] GetServices(IEnumerable<Type> services)
+                    if (value != null)
                     {
-                        classes.Add(aClass);
+                        var aClass = Class.FromType((Type)propertyInfo.GetValue(null, null));
```
use value here then

guperrot commented on SDK/MobileCenter/Microsoft.Azure.Mobile.iOS/MobileCenter.cs:
```diff
@@ -174,10 +174,14 @@ private static Class[] GetServices(IEnumerable<Type> services)
             var classes = new List<Class>();
             foreach (Type t in services)
```
var can be used here too

guperrot commented on SDK/MobileCenter/Microsoft.Azure.Mobile.iOS/MobileCenter.cs:
```diff
@@ -174,10 +174,14 @@ private static Class[] GetServices(IEnumerable<Type> services)
             {
-                if((GetBindingType(t) != null) && (GetClassForType(GetBindingType(t)) != null))
+                var bindingType = GetBindingType(t);
+                if((bindingType != null) && (GetClassForType(bindingType) != null))
```
I think `&& (GetClassForType(bindingType) != null)` can be removed as you check again right after

guperrot commented on SDK/MobileCenterDistribute/Microsoft.Azure.Mobile.Distribute.iOS/Distribute.cs:
```diff
@@ -7,8 +7,23 @@ namespace Microsoft.Azure.Mobile.Distribute
+        static Type _internalBindingType = typeof(iOSDistribute);
+
+        [System.Diagnostics.Conditional("DEBUG")]
+        static void SetInternalBindingTypeForDebug()
```
SetInternalBindingTypeForDebug => UnsetInternalBindingType

guperrot commented on SDK/MobileCenter/Microsoft.Azure.Mobile.Android/MobileCenter.cs:
```diff
@@ -184,7 +184,24 @@ private static Application SetWrapperSdkAndGetApplication()
         {
-            return services.Select(service => Class.FromType((Type)service.GetProperty("BindingType").GetValue(null, null))).ToArray();
+            var classes = new List<Class>();
+            foreach (Type t in services)
```
var here too

guperrot replied:
bump

TroubleMakerBen commented:
This approach does not work.

TroubleMakerBen commented:
Don't delete the branch as we will very likely do something here ;)

msftclas commented:
@TroubleMakerBen,
Thanks for your contribution as a Microsoft full-time employee or intern. You do not need to sign a CLA.
_Thanks,
Microsoft Pull Request Bot_

guperrot commented on SDK/MobileCenter/Microsoft.Azure.Mobile.iOS/MobileCenter.cs:
```diff
@@ -171,7 +171,20 @@ public static bool Enabled
         {
-            return services.Select(service => GetClassForType(GetBindingType(service))).ToArray();
+            var classes = new List<Class>();
+            foreach (Type t in services)
```
use var

guperrot commented on SDK/MobileCenterDistribute/Microsoft.Azure.Mobile.Distribute.iOS/Distribute.cs:
```diff
@@ -41,5 +49,21 @@ public static void OpenUrl(NSUrl url)
+
+        /// <summary>
+        /// Do not check for updates in case the app is launched with a debug configuration.
+        /// Place this method call into you app delegate's didFinishLaunching:withOptions: method BEFORE 
```
It's called FinishedLaunching in Xamarin.iOS

guperrot commented on SDK/MobileCenterDistribute/Microsoft.Azure.Mobile.Distribute.iOS/Distribute.cs:
```diff
@@ -41,5 +49,21 @@ public static void OpenUrl(NSUrl url)
+        /// <summary>
+        /// Do not check for updates in case the app is launched with a debug configuration.
+        /// Place this method call into you app delegate's didFinishLaunching:withOptions: method BEFORE 
+        /// you call MobileCenter.start(...) if you are using in-app updates.
```
start => Start

guperrot replied:
before you call MobileCenter.Start or bebore you init the forms application object if you use Xamarin Forms.

guperrot commented on SDK/MobileCenterDistribute/Microsoft.Azure.Mobile.Distribute.iOS/Distribute.cs:
```diff
@@ -41,5 +49,21 @@ public static void OpenUrl(NSUrl url)
+        /// </summary>
+        /// <remarks>
+        /// This method is required because the SDK cannot detect an attached debugger, nor can it detect
+        /// a release configuration. If this method is not called, the browser will appear and try to
```
nor can it detect release configuration at runtime.

guperrot commented on SDK/MobileCenterDistribute/Microsoft.Azure.Mobile.Distribute.iOS/Distribute.cs:
```diff
@@ -41,5 +49,21 @@ public static void OpenUrl(NSUrl url)
+        /// <remarks>
+        /// This method is required because the SDK cannot detect an attached debugger, nor can it detect
+        /// a release configuration. If this method is not called, the browser will appear and try to
+        /// setup in-app updates which we do not want.
```
I guess we can remove "which we do not want"

TroubleMakerBen commented:
Hey,
I'm talking to Ela in a few.
This also needs to be documented but there's a sep. PR for a readme update. Let's do that in a sep. PR.

TroubleMakerBen commented:
The method name has been approved by Ela and I addressed the comments.
The iOS SDK will get the same API this week, there will be a sep. PR for it later so the Xamarin SDK doesn't have do the actual logic

guperrot commented on SDK/MobileCenterDistribute/Microsoft.Azure.Mobile.Distribute.iOS/Distribute.cs:
```diff
@@ -41,5 +49,22 @@ public static void OpenUrl(NSUrl url)
+        /// </summary>
+        /// <remarks>
+        /// This method is required because the SDK cannot detect an attached debugger, nor can it detect
+        /// a release configuration at runtme. If this method is not called, the browser will appear and try to
```
typo in runtime

TroubleMakerBen commented:
I fixed the typo.
